### ISSUE TITLE
Unity and JPEG Compression

### ISSUE THREAD

**Issue by NPatch** [Microsoft/RoomAliveToolkit#82]
The tutorial notes that JPEG compression should be enabled. I understand that it's better for KinectServers exchanging data since it's a compressed package of data. But Unity doesn't play as well with it.
Even though I've known this from experience, I see a 3-4x speedup in framerate when I turn it off. Every frame that the server actually gives new data, Tex2D.LoadImage takes up to 40ms which acts like a spike every so often. When I turn JPEG compression off, the frame time goes down to 11-12ms avg with some spikes at 20ms. At this point I'm trying to figure out how to do a 1x1 setup but eventually it'll have to scale up to 3x3. Will this be an issue even if switches are used? Like eliminating the wifi connectivity issues?

**Comment by NPatch** [comment]
Any ideas on this one?!

**Comment by NPatch** [comment]
By the way, if you disable JPEG compression, Unity does not know how to handle the input and renders a big white texture with a red question mark. I tried adding some code in RATKinectClient to use LoadTextureRawData when uncompressed Color is streamed but the type of the image data is 0. So it's not easy to understand whether it's BGRA,RGBA or whatever else Kinect can give. Also tried toggling Stream RAW Color and Process Color Frames. In both cases, the texture viz is wrong. In one case you get the question mark and in the other you get a big gray image(possible alpha).

**Comment by NPatch** [comment]
My bad...I forgot to call Apply to upload the loaded buffer to the gpu,after LoadRawTextureData. LoadImage does that automatically. Now I have color info, but not correct.

**Comment by NPatch** [comment]
I was talking about the Unity version of KinectServer though. If JPEG compression is turned on, it might be faster to transmit but it's a lot slower to load into a Texture in the Unity runtime. On the other hand, if you turn the compression off, the texture in Unity gets corrupted because the code was written to only work with JPEG compression on.

**Comment by thundercarrot** [comment]
I'm talking about the Unity version of KinectServer as well. You might need a faster computer? Also, do you really need the color information?

**Comment by NPatch** [comment]
Btw, this might not be the appropriate thread, but I have a question. How does the calibration handle Kinects that are not covering projections if at all? So for mirror like scenarios, there would be a need for an extra sensor that is above the projection looking down. But that sensor, if in a cave envirnoment will either capture parts of graycodes from the side projections, otherwise nothing at all. Does it get ignored if it captures none of the projections? If it can take part(for some reason I can't think of currently), can it be the primary coordinate system where all other sensors get converted to? Or would we have to provide pose matrices etc, by hand to support at the very least the conversion to the primary coordinate system provided by one of the valid Kinects in the setup?

**Comment by thundercarrot** [comment]
The first camera listed in the calibration file establishes the coordinate system. All other camera and projector poses are in that first cameras coordinate system. If you were to successfully calibrated, the pose for the camera would be the identity matrix.

If you have another camera that doesn't participate in the calibration, I'm not sure how RAT behaves; I would not be suprised if it barfs. In which case it is best to treat it separately and find some other means of calibrating it. E.g., this setup:

https://www.youtube.com/watch?v=9A18AxfC2tM&t=0s&list=PLKcZe1FzCK3i-TjlWzP9TqBD6rGo1FWi6&index=14

I added a fourth camera for body tracking and established its pose wrt to camera 0 by hand.

I gotta admit I'm a little flat footed on the Unity JPEG decompression issue, as I have been lately working on a new release of RoomAlive Toolkit that pushes most of the heavy lifting into a native plugin for Unity. This approach allows for much more modularity and re-use when moving from one game engine or rendering framework to another. This currently supports 8 Kinect cameras at video rate. For JPEG decompression it uses the Windows built-in WIC codec which is quite fast. It also uses an entirely new REST server that allows for frames to be dropped.

I'm puzzled as to why uncompressed color would work at all given the bandwidth calculations I included earlier.

**Comment by NPatch** [comment]
Ok, thanks for answering. I was already aware of that video and I think you mentioned it in a Channel 9 post that the face tracking Kinect is separately treated but that post/video was from 2015 and I was wondering whether that had changed at all. I'm still hoping that if I place a Kinect above the Center wall projection looking down at the user, it might work if it can at least see some of the gray codes from the side walls.

Yeah, this issue has been a bane for me. Same goes for the PNG loading. For the record,I'm using Unity 5.6.6, so I'm not aware of any speedups in 20XX versions. As for the native plugin, I assume you mean a rendering plugin that receives data directly from the sensors in C++ land and updates DX11 texture resources and returning those DX11 texture resource pointers back to Unity so the C# Texture2D is basically just a wrapper around the DX11 pointer, you're managing yourself. I've written one before as a perf experiment to compare with the classic Kinect v2 SDK unity plugin and it did quite well. Also the ability to do optimizations on data layouts of other data(like joints) was very welcome. I'm sure it's nowhere near what you're making though. Will it be available publicly? Or is it a pet project?

I'm not sure what decoders Unity uses, but decoding expands the data to the same size as the uncompressed before commiting the data to the gpu and its cpu counterpart. So higher loading times for compressed data should make sense since it's extra work. Right?

**Comment by thundercarrot** [comment]
Yes, I plan to release it. I can't make any promises as to when though.

Yes, it makes sense that you would have higher load times for compressed data. You can trade that off the delay in network transfer for when you do not compress. I speak to this a bit in a paper I published last year on depth image compression:
https://dl.acm.org/citation.cfm?id=3134144

**Comment by thundercarrot** [comment]
In our experience the NUCs work very well here. Re: other laptops etc. keep in mind that the Kinect can be very picky about what USB3 chipset is used. This is documented elsewhere. The NUCs from Intel use the Intel chipset (unsurprisingly) which is one of the recommended makes.

**Comment by NPatch** [comment]
I know, as I said in a previous response, the other day I changed from an Alienware 14 laptop because all sensors I used, in all of its 3 USB3.0 ports, had a ~1000ms response time in the calibration version of the KinectServer and as soon as I switched to the desktop, I got ~100ms. Weird part is my Lenovo Y50-70 laptop which was the machine of the second Kinect in the setup, which sent packets over a network switch had ~250ms response time and was way faster in delivering the depth images during calibration than the Kinect physically connected to the machine running the calibration. On paper, both laptops should have the same Intel chipset, but when I ran the Intel Chipset Identification Utility, Alienware came up as Mobile Express while my laptop came up as Intel 8 series( on lenovo spec it was HM87 and I think the same one is mentioned in the Alienware 14 spec sheet). Obviously one of the two is different somehow, but I haven't figured out why yet. 

I was also referring to NUC's gpu and not just the USB3/PCI part of the process. Could you share the NUC models? If you aren't doing anything custom to it, it would be incredibly helpful to know of a small machine that can satisfy needs, for NxN scenarios. Presumably it would minimize costs for all machines other than the primary one which has lots of requirements(calibration hub, Kinect, all projectors etc) and has to be heavy duty.

As for the documentation, it also says Renesas but I've come across posts that say Renesas didn't work. Compatibility is an issue that's still coming up in new MSDN posts. I'm guessing it's also an arch issue, otherwise we'd have more concrete info on how to choose external PCIs by now.

**Comment by tkbala** [comment]
I realize that this is an old thread. But it turns out I am facing the same issue. Any thoughts now on how one could speed up the "Tex2D.LoadImage" for loading the RGB frame data?

**Comment by NPatch** [comment]
When I tested things, I was using Unity 5.6 if I'm not mistaken. I don't know if 20XX versions have touched up on the jpeg de/encoder. That's the only thing I can think of at this point.

**Comment by tkbala** [comment]
I am using the 2018.3 version and running into the same issue. I was looking for ways to run the jpeg decoder on a parallel thread but I haven't had any success doing that yet

**Comment by NPatch** [comment]
AFAIK, the regular API in Unity is not thread safe so you can't call it outside the main thread. I guess the only way to do it would be in a native render plugin where you create your own D3DTexture2D, use your own decompressor/decoder and give Unity a pointer to it to use as a Texture2D object. If Unity 2019 hasn't introduced a change that allows you to do it in C#, going C++ is the only way. Can't help with that though.

**Comment by nsmith1024** [comment]
Has anybody solved this issue? Im having really bad performance and stuttering when loading images. I think a solution is to let the browser load it for you, similar to how the WebGLMovieTexture works which uses the browser to play videos on surfaces within the  unity scene. WebGLMovieTexture is located for free on the asset store, you just pass it the URL of the movie you want played and the surface you want to play it on, and it plays it asynchronously without any stuttering.

Maybe one way to speed up LoadImages is to do the same thing somehow using images rather than a video, pass the URL of the image to the browser via javascript, then let it draw the image on to the given unity object asynchronously.
